EscrowFactory using CREATE2 opcode for deterministic address generation

lipmas opened the pull request.

lipmas committed:
[Dependencies] Update npm solc to v0.5.12

lipmas committed:
[Escrow, Library, Factory] Use CREATE2 for deterministic addresses

+ Use CREATE2 to create escrows at a deterministic address using the
escrowParamsHash as salt and hard coded escrow contract bytecode. This
eliminates issues with relying on the funding address of the escrow
not changing once the escrow params have been agreed on.

+ Simplify logic for opening escrows since we can rely on the escrow
address being prefunded before the call to the factory that creates
the escrow. Therefore the create function on the factory can just
trigger the opening logic in the EscrowLibrary when a new escrow is
being created. This removes the need for a checkFunded function or a
EscrowFunded/EscrowOpened events. Finally the EthEscrow payable
fallback can be removed because now its assumed all funds will be sent
to the address prior to its creation.

+ Remove the escrow params mapping in the factory as now the create2
call will fail if a duplicate escrow is created.

lipmas committed:
[Tests] Fix tests to work with new determinstic address

+ Tests now precompute escrow addresses from the escrowParams and fund
the addresses proior to calling the factory to create the contract.

lipmas committed:
[Solc] Adds an external truffle compile script

compile-constracts.sh is used to standardize the compiler
version/flags and metadata that is generated when compiling some of
the smart contracts. This is needed for EthEscrow.sol and
ERC20Escrow.sol because we hard-code the bytecode of these conracts in
the EscrowFactory in order to use the create2 opcode.

lipmas committed:
Add EscrowFactoryWithErc20.sol to decouple from EscrowFactory.sol

EthanHeilman commented:
>"Simplify logic for opening escrows since we can rely on the escrow address being prefunded before the call to the factory that creates the escrow. Therefore the create function on the factory can just trigger the opening logic in the EscrowLibrary when a new escrow is being created. This removes the need for a checkFunded function or a EscrowFunded/EscrowOpened events."

What happens if:
1. Alice funds the escrow contract address,
2. then she opens 

but her funding transaction isn't confirmed or is confirmed but after the escrow contract is created?

Nabeelperson commented:
@EthanHeilman Since both of these transactions will be associated with a nonce from Alice's account, as long as she performs the transactions in the appropriate order, there should always be funds committed to the address before funding.

EthanHeilman commented:
@Nabeelperson In Arwen the account which sends the message to the create the escrow contract is not the same as the account which funds the escrow. No ordering is enforced.

EthanHeilman commented on contracts/EscrowFactory.sol:
```diff
@@ -48,14 +52,11 @@ contract EscrowFactory is Ownable {
-        EthEscrow escrow = new EthEscrow(
-            address(escrowLibrary)
-        );
+        bytes memory bytecode = abi.encodePacked(ETH_ESCROW_BYTECODE, abi.encode(address(escrowLibrary)));
```
What does appending the address of the escrowLibrary to the bytecode do?

lipmas replied:
When the Escrow contract is created its code will initialized with the bytecode and then the rest of the data will be passed to the constructor of the escrow (which takes the escrowLibrary as a parameter).

https://github.com/cwcrypto/arwen-eth-contracts/blob/c872e969ec507b00e8a41b792867c412f2344833/contracts/Escrow.sol#L25

lipmas committed:
[EscrowFactory] Remove `selfDestruct` from the EscrowFactory

cwhinfrey commented on contracts/EscrowFactoryWithERC20.sol:
```diff
@@ -0,0 +1,51 @@
+    public
+    {
+        bytes32 escrowParamsHash = keccak256(abi.encodePacked(
+            address(this),
```
`address(this)` isn't necessary here because it's already factored into the address for the create2 deployment. It also doesn't hurt though.
create2 address calculation: `keccak256( 0xff ++ address ++ salt ++ keccak256(init_code))`

cwhinfrey replied:
Same with the `escrowParamsHash` calculation in `EscrowFactory`

lipmas replied:
Yes thats true its not necessary as part of the salt for create2. Although we still emit this value as an indexed parameter in the `EscrowCreated` event so its nice to make sure this value is guaranteed to be unique even across different escrow factory deployments.

lipmas committed:
[EscrowFactory] Use type().creationCode to reference escrow bytecode instead of hard-coding

lipmas committed:
[Solc] Remove external compiler from truffle config

Now test environments using different compiler version/metadata dont
cause issues because the escrow contract bytecode is no longer
hard-coded into the EscrowFactory.

lipmas committed:
[EscrowFactory] Fix solhint maximum line length error

EthanHeilman reviewed (approved):
LGTM
